### PR TITLE
8326666: Remove the Java Management Extension (JMX) Subject Delegation feature

### PR DESCRIPTION
**Pull request by kevinjwalls**
The deprecated Subject Delegation feature in JMX will be removed.

This was marked in JDK 21 as deprecated for removal (JDK-8298966).

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change requires CSR request [JDK-8326691](https://bugs.openjdk.org/browse/JDK-8326691) to be approved

### Issues
 * [JDK-8326666](https://bugs.openjdk.org/browse/JDK-8326666): Remove the Java Management Extension (JMX) Subject Delegation feature (**Enhancement** - P3)
 * [JDK-8326691](https://bugs.openjdk.org/browse/JDK-8326691): Remove the implementation of JMX Subject Delegation and degrade the applicable APIs to throw UnsupportedOperationException (**CSR**)

### Reviewers
 * [Mandy Chung](https://openjdk.org/census#mchung) (@mlchung - **Reviewer**) ⚠️ Review applies to [91ec015f](https://git.openjdk.org/jdk/pull/18025/files/91ec015f73ed5cda9cc8fd93c31b3463a941daad)
 * [Daniel Fuchs](https://openjdk.org/census#dfuchs) (@dfuch - **Reviewer**) ⚠️ Review applies to [43b10a12](https://git.openjdk.org/jdk/pull/18025/files/43b10a126dc611696ffe3a2564c36e4e76ff89e1)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/18025/head:pull/18025` \
`$ git checkout pull/18025`

Update a local copy of the PR: \
`$ git checkout pull/18025` \
`$ git pull https://git.openjdk.org/jdk.git pull/18025/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 18025`

View PR using the GUI difftool: \
`$ git pr show -t 18025`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/18025.diff">https://git.openjdk.org/jdk/pull/18025.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/18025#issuecomment-1972853497)

### CODE REVIEW

**Comment by kevinjwalls** [comment]
This covers the public interface to the feature, and tests.

Locally the feature is refused during a call to getMBeanServerConnection(with a non-null delegationSubject).

Remotely, `javax.management.remote.rmi.RMIConnectionImpl` has methods which accept a Subject parameter for Subject Delegation.  This cannot now be non-null in local usage, but an RMI Connection from an older JDK can invoke these methods with a non-null delegationSubject.  Such access will be recognised and an UnsupportedOperationException will be thrown (which the client will see).

RMIConnectionImpl can be simplified, and the class SubjectDelegator removed.

JMXSubjectDomainCombiner I would like to remove separately after this is done.

File: src/java.management/share/classes/javax/management/remote/JMXConnector.java [hunk 1]
```diff
@@ -175,9 +155,14 @@ public MBeanServerConnection getMBeanServerConnection()
      * deprecated and subject to removal. There is no replacement.
      */
     @Deprecated(since="21", forRemoval=true)
-    public MBeanServerConnection getMBeanServerConnection(
-                                               Subject delegationSubject)
-            throws IOException;
+    public default MBeanServerConnection getMBeanServerConnection(Subject delegationSubject)
```
Comment (AlanBateman) [inline src/java.management/share/classes/javax/management/remote/JMXConnector.java:158]: Specifying the UOE in the param description is problematic. I think you want to specify delegationSubject as not used, must be null, then declare UOE as one of the exceptions thrown by the method.

File: src/java.management.rmi/share/classes/javax/management/remote/rmi/RMIConnection.java [hunk 2]
```diff
@@ -123,9 +123,7 @@ public interface RMIConnection extends Closeable, Remote {
      *
      * @param className The class name of the MBean to be instantiated.
      * @param name The object name of the MBean. May be null.
-     * @param delegationSubject The <code>Subject</code> containing the
-     * delegation principals or <code>null</code> if the authentication
-     * principal is used instead.
+     * @param delegationSubject No longer supported and should be null.
```
Comment (AlanBateman) [inline src/java.management.rmi/share/classes/javax/management/remote/rmi/RMIConnection.java:126]: I think this interface will require all methods that take a delegation subject to specify UOE then the subject is not null.

**Comment by dfuch** [comment]
In addition to Alan's remarks I believe we need new tests to verify the new behavior.

File: src/java.management/share/classes/javax/management/remote/JMXConnector.java (hunk 1 above)
Comment (kevinjwalls) [reply src/java.management/share/classes/javax/management/remote/JMXConnector.java:161]: Thanks,  updated.

**Comment by kevinjwalls** [comment]
Added a test to ensure  jmxc.getMBeanServerConnection(delegationSubject) throws UnsupportedOperationException.

**Comment by seanjmullan** [comment]
Is there any value in keeping `SubjectDelegationPermission` after this change? If so, I would mark that API deprecated for removal, so that it can be removed in the next release or two.

**Comment by seanjmullan** [comment]
Are there remaining tests that test the JMX fine grained permissions feature w/o depending on subject delegation?

File: src/java.management/share/classes/javax/management/remote/JMXConnector.java [hunk 3]
```diff
@@ -135,49 +135,37 @@ public MBeanServerConnection getMBeanServerConnection()
...
-     * of a given delegation subject.
+     * interface by forwarding its methods to the remote MBean server.
      *
      * @exception IOException if a valid <code>MBeanServerConnection</code>
      * cannot be created, for instance because the connection to the remote
      * MBean server has not yet been established (with the {@link #connect(Map)
      * connect} method), or it has been closed, or it has broken.
      *
+     * @exception UnsupportedOperationException if a non-null delegationSubject
+     * is specifed. Subject Delegation has been removed.
```
[the remark this answers is not in the document]
Comment (kevinjwalls) [reply src/java.management/share/classes/javax/management/remote/JMXConnector.java]: got it thanks.

File: src/java.management/share/classes/javax/management/remote/JMXConnector.java [hunk 4]
```diff
@@ -135,49 +135,37 @@ public MBeanServerConnection getMBeanServerConnection()
...
+     *
      * @deprecated This method supported the legacy Subject Delegation feature,
      * and is only useful in conjunction with other APIs which are deprecated and
      * subject to removal in a future release. Consequently, this method is also
      * deprecated and subject to removal. There is no replacement.
```
[the remark this answers is not in the document]
Comment (kevinjwalls) [reply src/java.management/share/classes/javax/management/remote/JMXConnector.java]: Yes, that should be updated, the feature is gone with this change.

File: src/java.management/share/classes/com/sun/jmx/remote/internal/ClientNotifForwarder.java [hunk 5]
```diff
@@ -1,5 +1,5 @@
 /*
- * Copyright (c) 2002, 2022, Oracle and/or its affiliates. All rights reserved.
+ * Copyright (c) 2002, 2024 Oracle and/or its affiliates. All rights reserved.
```
[the remark this answers is not in the document]
Comment (kevinjwalls) [reply src/java.management/share/classes/com/sun/jmx/remote/internal/ClientNotifForwarder.java]: > Missing `,` after 2024.

oops, thanks!

**Comment by dfuch** [comment]
> Is there any value in keeping `SubjectDelegationPermission` after this change? If so, I would mark that API deprecated for removal, so that it can be removed in the next release or two.

No issue with deprecation. I guess it can be removed once `java.security.Policy` is removed?

File: src/java.management.rmi/share/classes/javax/management/remote/rmi/RMIConnector.java [hunk 6]
```diff
@@ -623,14 +606,15 @@ private Integer[] addListenersWithSubjects(ObjectName[]       names,
...
         public RemoteMBeanServerConnection() {
             this(null);
         }
 
         public RemoteMBeanServerConnection(Subject delegationSubject) {
-            this.delegationSubject = delegationSubject;
+            if (delegationSubject != null) {
+                throw new SecurityException("Subject Delegation has been removed.");
+            }
         }
```
Comment (dfuch) [inline src/java.management.rmi/share/classes/javax/management/remote/rmi/RMIConnector.java:618]: Can't this constructor be removed? It's a private inner class so we should be able to throw before calling that constructor?

File: test/jdk/javax/management/remote/mandatory/notif/DeadListenerTest.java [hunk 7]
```diff
@@ -75,11 +75,11 @@ public static void main(String[] args) throws Exception {
         RMIConnectorServer cs = new RMIConnectorServer(url, null, rmiServer, mbs);
         cs.start();
         JMXServiceURL addr = cs.getAddress();
-        assertTrue("No connections in new connector server", rmiServer.connections.isEmpty());
+        assertTrue("Expected no connections in new connector server", rmiServer.connections.isEmpty());
```
Comment (dfuch) [inline test/jdk/javax/management/remote/mandatory/notif/DeadListenerTest.java:78]: Changes to this file seem unrelated to the removal of the feature - could you confirm that they were intended (no issue if they were intended).

File: src/java.management.rmi/share/classes/javax/management/remote/rmi/RMIConnection.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (dfuch) [reply src/java.management.rmi/share/classes/javax/management/remote/rmi/RMIConnection.java:126]: I was wondering if we should add a note at the class-level API documentation to say that subject delegation has been removed, and that methods will throw UOE if a non-null delegation subject is passed.

[the remark this answers is not in the document]
Comment (dfuch) [reply src/java.management.rmi/share/classes/javax/management/remote/rmi/RMIConnection.java:126]: Maybe mention there that the delegationSubject parameter is kept for interoperability with older clients.

[the remark this answers is not in the document]
Comment (mlchung) [reply src/java.management.rmi/share/classes/javax/management/remote/rmi/RMIConnection.java:126]: It's a good suggestion.   Something like this:

All methods that take a `Subject` parameter will throw UOE if called with a non-null subject delegation.  JMX subject delegation feature is no longer supported.

**Comment by seanjmullan** [comment]
> > Is there any value in keeping `SubjectDelegationPermission` after this change? If so, I would mark that API deprecated for removal, so that it can be removed in the next release or two.
> 
> No issue with deprecation. I guess it can be removed once `java.security.Policy` is removed?

Yes, or to be more precise, when the `JavaPolicy` implementation of `java.security.Policy` is removed.

File: test/jdk/javax/management/remote/mandatory/notif/DeadListenerTest.java (hunk 7 above)
Comment (kevinjwalls) [reply test/jdk/javax/management/remote/mandatory/notif/DeadListenerTest.java:78]: Yes, they were intentional, should have made that clear.  I had a failure from this test at some point, and the assert message made no sense to me.
e.g. "New value of count1 == 1" is shown, but that's what was expected and not what was seen.

File: src/java.management.rmi/share/classes/javax/management/remote/rmi/RMIConnector.java (hunk 6 above)
Comment (kevinjwalls) [reply src/java.management.rmi/share/classes/javax/management/remote/rmi/RMIConnector.java:618]: Yes that can disappear with no impact.

File: src/java.management/share/classes/javax/management/remote/JMXConnector.java [hunk 8]
```diff
@@ -135,49 +135,35 @@ public MBeanServerConnection getMBeanServerConnection()
...
 
     /**
-     * <p>Returns an <code>MBeanServerConnection</code> object representing
-     * a remote MBean server on which operations are performed on behalf of
-     * the supplied delegation subject. For a given <code>JMXConnector</code>
-     * and <code>Subject</code>, two successful calls to this method will
-     * usually return the same <code>MBeanServerConnection</code> object,
-     * though this is not required.</p>
+     * <p>This method remains for compatibility reasons, but has no more meaning
+     * than {@link #getMBeanServerConnection()}.
```
[the remark this answers is not in the document]
Comment (kevinjwalls) [reply src/java.management/share/classes/javax/management/remote/JMXConnector.java:139]: ok

File: src/java.management.rmi/share/classes/javax/management/remote/rmi/RMIConnection.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (kevinjwalls) [reply src/java.management.rmi/share/classes/javax/management/remote/rmi/RMIConnection.java:126]: OK sure -- didn't want to make the individual delegationSubject mentions too terse, but if there is some earlier text then that could be OK...

[the remark this answers is not in the document]
Comment (kevinjwalls) [reply src/java.management.rmi/share/classes/javax/management/remote/rmi/RMIConnection.java]: Thanks, updating with those changes. Additionally, in JMXConnector I'm removing the text "Calling this method is equivalent to calling getMBSC(null)..." as that is pointless, and possibly misleading now.

File: src/java.management.rmi/share/classes/javax/management/remote/rmi/RMIConnection.java [hunk 9]
```diff
@@ -81,6 +81,11 @@
  * of a method is not specified here, it is the same as in the
  * corresponding <code>MBeanServerConnection</code> method.
  *
+ * <p>JMX Subject Delegation has been removed. All methods that take a
+ * delegationSubject parameter will throw UnsupportedOperationException if
```
Comment (mlchung) [inline src/java.management.rmi/share/classes/javax/management/remote/rmi/RMIConnection.java:85]:
```suggestion
 * <p>JMX Subject Delegation has been removed. All methods that take a
 * delegation subject parameter will throw {@code UnsupportedOperationException} if
```

**Review by mlchung — APPROVED** [review]
The spec change looks good to me.    I leave to others to review the implementation and test changes.

**Comment by kevinjwalls** [comment]
> The spec change looks good to me. I leave to others to review the implementation and test changes.

Thanks!

**Comment by kevinjwalls** [comment]
Thanks @dfuch for the comments and review. 8-)

**Review by mlchung — APPROVED** [review]
The spec updates are good with me.   Caught one formatting nit in the `RMIConnection` class spec.

**Review by dfuch — APPROVED** [review]
